app/gosqs/gosqs.go

```diff
@@ -114,9 +115,42 @@ func ListQueues(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
+type CreateQueueRequest struct {
+	QueueName  string            `json: QueueName`
+	Attributes map[string]string `json: Attributes`
+	Tags       map[string]string `json: Tags`
```

Comment:
TLDR: #2 is the better solution but more work than I wanted to take on.

Generally I agree with you on #2 being the better solution, it's just a lot more work so I didn't take it on and did not feel comfortable refactoring the codebase to use a shared internal type.

With all these new proposed API types, you will need custom marshallers to transform the API Types (JSON / AWS Query Protocol) to the new internal type, then update all the internal code to use the new type, and tests as well.

Also, due to the current design having every method be its own handler with signature:
```
FuncName(w http.ResponseWriter, req *http.Request)
```

You'll need some sort of repetitive if/else to check in each handler to parse the amz/json header and then parse the incoming payload differently each time (like what kojisaiki ended up doing with parseCreateQueueRequestBody()).

This all seemed like a lot of work so I didn't take that on 😅 . 

If you want to get very clean, you could break out the JSON / XML **response** types as well and marshall from your internal types to the json or xml response types. What I have in place in my branch works (and took a lot of [referencing of the docs](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html#API_ReceiveMessage_ResponseSyntax), but has some duplicated fields that are not overlapping between json/xml b/c the format of MessageAttributes (along with some other field names) changed with the new API version.

For me -- I burned my whole morning and early afternoon getting what's in my PR working so I could unblock myself and keep moving forward. I unfortunately don't have time for the scale of the proposed refactor above, or, at least, wouldn't be able to complete it any a timely fashion.

I love your library, and would like to thank you for writing it 👍 and for responding here as well!

Reply:
No sweat - happy you got something working for you.  I will likely PR onto this or create an alternative for the full feature set then when I have time.  You are right it will be a lot of work but I think it could be done endpoint by endpoint, which is what I'll try to do.  Feel free to review the future changes if you are interested!  Thanks again for putting this out there!

Reply:
^ also me, btw - in case anyone is wondering.  🙃